File: CsLua/Collection/CsLuaList.lua
--TargetFile: CsLua.lua

CsLua = CsLua or {};
CsLua.Collection = CsLua.Collection or {};
CsLua.Collection.CsLuaList = function(generics)
	local class = {};
	local getters = {};
	local publicClass = {};

	local t = {};

	local Cstor = function(...)
		local args = {...};
		if type(args[1]) == "table" then
			for _,v in ipairs(args[1]) do
				publicClass.Add(v);
			end
		end
	end

	local Initialize = function(nativeTable)
		t[0] = nativeTable[0];
		for i, v in ipairs(nativeTable) do
			t[i] = v;   
		end
	end

	local Serialize = function(info)
		for i, v in pairs(t) do
			info.AddValue(publicClass, i, v, generics);
		end
	end

	local Deserialize = function(info)
		for _, key in ipairs(info.GetValueKeys(publicClass)) do
			t[key] = info.GetValue(publicClass, key, generics);
		end
	end

	CsLua.CreateSimpleClass(class, publicClass, "CsLuaList", "CsLua.Collection.CsLuaList", Cstor, Initialize, Serialize, Deserialize,  {{"System.Collection.Generic.IList"}});

	getters.Count = function()
		local c = #(t);
		if t[0] then
			c = c + 1;
		end
		return c;
	end

	class.GetEnumerator = function()
		return function(_, prevKey) 
			if prevKey == nil then
				if t[0] then
					return 0, t[0];
				end
				if t[1] then
					return 1, t[1];
				end
				return nil, nil;
			end
			local k = prevKey + 1;
			if t[k] then
				return k, t[k];
			end
			return nil, nil;
		end;
	end

	class.Add = function(value)
		local c = publicClass.Count;
		publicClass[c] = value;
		return c;
	end

	class.Clear = function()
		t = {};
	end

	class.Contains = function(value)
		for i = 0, publicClass.Count - 1 do
			if publicClass[i] == value then
				return true;
			end
		end
		return false;
	end

	class.IndexOf = function(value)
		for i = 0, publicClass.Count - 1 do
			if publicClass[i] == value then
				return i;
			end
		end
		return -1;
	end

	class.Insert = function(index, value)
		if (index == 0) then
			table.insert(t, 1, t[0]);
			t[0] = value;
		else
			table.insert(t, index, value);
		end
	end

	class.RemoveAt = function(index)
		if index > 0 then
			table.remove(t, index);
		elseif index == 0 then
			t[0] = table.remove(t, 1)
		end
	end

	class.Remove = function(value)
		publicClass.RemoveAt(publicClass.IndexOf(value));
	end

	class.ToNativeLuaTable = function()
		local native = {};
		for i = 0, publicClass.Count - 1 do
			native[i+1] = publicClass[i];
		end
		return native;
	end

	-- Linq support
	class.Where = function(f)
		local t = CsLua.Collection.CsLuaList();
		for i = 0, publicClass.Count - 1 do
			if f(publicClass[i]) then
				t.Add(publicClass[i]);
			end
		end
		return t;
	end

	class.Select = function(f)
		local t = CsLua.Collection.CsLuaList();
		for i = 0, publicClass.Count - 1 do
			t.Add(f(publicClass[i]));
		end
		return t;
	end

	class.Any = function(f)
		if f then
			return publicClass.Where(f).Count > 0;
		else
			return publicClass.Count > 0;
		end
	end

	class.FirstOrDefault = function(f)
		if f then
			return publicClass.Where(f)[0];
		else
			return publicClass[0];
		end
	end

	class.First = function(f)
		local res = publicClass.FirstOrDefault(f);
		if res then
			return res;
		end
		CsLuaMeta.Throw(CsLua.CsException().__Cstor("The source sequence is empty."));
	end

	class.Single = function(f)
		if publicClass.Count == 1 then
			return publicClass[0];
		end
		CsLuaMeta.Throw(CsLua.CsException().__Cstor("The source sequence did not hold precisely one element."));
	end

	class.SingleOrDefault = function(f)
		if publicClass.Count == 1 then
			return publicClass[0];
		elseif publicClass.Count == 0 then
			return nil;
		end
		CsLuaMeta.Throw(CsLua.CsException().__Cstor("The source sequence did not hold one or none elements."));
	end

	class.LastOrDefault = function(f)
		if f then
			local t = publicClass.Where(f);
			return t[t.Count - 1];
		else
			return publicClass[publicClass.Count - 1];
		end
	end

	class.Last = function(f)
		local res = publicClass.LastOrDefault(f);
		if res then
			return res;
		end
		CsLuaMeta.Throw(CsLua.CsException().__Cstor("The source sequence is empty."));
	end

	class.Union = function(otherList)
		local res = CsLua.Collection.CsLuaList();
		for i = 0, publicClass.Count - 1 do
			res.Add(publicClass[i]);
		end
		for i = 0, otherList.Count - 1 do
			res.Add(otherList[i]);
		end
		return res;
	end

	class.Distinct = function()
		local res = CsLua.Collection.CsLuaList();
		for i = 0, publicClass.Count - 1 do
			local v1 = publicClass[i]
			local found = false;
			for j = 0, res.Count - 1 do
				local v2 = publicClass[j];
				if v1 == v2 then
					found = true;
					break;
				end
			end
			if found == false then
				res.Add(v1);
			end
		end
	end

	class.ToList = function()
		return publicClass.Where(function() return true; end);
	end

	class.OrderBy = function(f)
		local t = publicClass.ToNativeLuaTable();
		table.sort(t, function(a,b) return f(a) < f(b) end);
		return CsLua.Collection.CsLuaList().__Cstor(t);
	end

	class.Skip = function(n)
		local l = CsLua.Collection.CsLuaList();
		for i = n, publicClass.Count - 1 do
			l.Add(publicClass[i]);
		end
		return l;
	end

	class.Take = function(n)
		local l = CsLua.Collection.CsLuaList();
		for i = 0, n - 1 do
			l.Add(publicClass[i]);
		end
		return l;
	end

	class.Foreach = function(action)
		for i = 0, publicClass.Count - 1 do
			action(publicClass[i]);
		end
	end

	class.Max = function(selector)
		local c = nil;
		for i = 0, publicClass.Count - 1 do
			local v = selector(publicClass[i]);
			c = math.max(c or v, v);
		end
		return c;
	end

	class.Min = function(selector)
		local c = 0;
		for i = 0, publicClass.Count - 1 do
			local v = selector(publicClass[i]);
			c = math.min(c or v, v);
		end
		return c;
	end

	class.Sum = function(selector)
		local c = 0;
		for i = 0, publicClass.Count - 1 do
			c = c + selector(publicClass[i]);
		end
		return c;
	end

	setmetatable(publicClass, {
		__index = function(_, key)
			if type(key) == "number" then
				return t[key];
			end
			if getters[key] then
				return getters[key]();
			end
			if class[key] then
				return class[key]
			end
		end,
		__newindex = function(_, key, value)
			if type(key) == "number" then
				t[key] = value;
			end
		end,
	});

	return publicClass;    
end

File: CsLua/Collection/CsLuaList.cs
namespace CsLua.Collection
{
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Linq;
    using Lua;
    using System.Runtime.Serialization;

    [Serializable]
    public class CsLuaList<T> : IList<T>, ISerializable
    {
        protected IList<T> list;

        public CsLuaList()
        {
            this.list = new List<T>();
        }

        public CsLuaList(SerializationInfo info, StreamingContext context)
        {
            this.list = new List<T>();
            var c = (int)info.GetValue("__size", typeof(int));
            for (var i = 0; i < c; i++)
            {
                var pair = (KeyValuePair<object, object>)info.GetValue(i + "", typeof(KeyValuePair<object, object>));
                this.list.Add((T)pair.Value);
            }
        }

        protected CsLuaList(NativeLuaTable table)
        {
            throw new NotImplementedException();
        }

        protected CsLuaList(IList<T> innerList)
        {
            this.list = innerList;
        }

        public int IndexOf(T item)
        {
            return this.list.IndexOf(item);
        }

        public void Insert(int index, T item)
        {
            this.list.Insert(index, item);
        }

        public void RemoveAt(int index)
        {
            this.list.RemoveAt(index);
        }

        public T this[int index]
        {
            get
            {
                return this.list[index];
            }
            set
            {
                this.list[index] = value;
            }
        }

        public void Add(T value)
        {
            this.list.Add(value);
        }

        public void Clear()
        {
            this.list.Clear();
        }

        public bool Contains(T item)
        {
            return this.list.Contains(item);
        }

        public void CopyTo(T[] array, int arrayIndex)
        {
            this.list.CopyTo(array, arrayIndex);
        }

        public bool Remove(T item)
        {
            return this.list.Remove(item);
        }

        public int Count
        {
            get { return this.list.Count; }
        }

        public bool IsReadOnly { get; private set; }

        public IEnumerator<T> GetEnumerator()
        {
            return this.list.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.list.GetEnumerator();
        }

        public NativeLuaTable ToNativeLuaTable()
        {
            var t = new NativeLuaTable();
            for (var i = 0; i < this.list.Count; i++)
            {
                t[i] = this.list[i];
            }
            return t;
        }

        public CsLuaList<T> Take(int i)
        {
            return new CsLuaList<T>(this.list.Take(i).ToList());
        }

        public CsLuaList<T> Skip(int i)
        {
            return new CsLuaList<T>(this.list.Skip(i).ToList());
        }

        public CsLuaList<T> Where(Func<T, bool> condition)
        {
            return new CsLuaList<T>(this.list.Where(condition).ToList());
        }

        public CsLuaList<TResult> Select<TResult>(Func<T, TResult> selector)
        {
            return new CsLuaList<TResult>(this.list.Select(selector).ToList());
        }

        public bool Any()
        {
            return this.list.Any();
        }

        public bool Any(Func<T, bool> condition)
        {
            return this.list.Any(condition);
        }

        public CsLuaList<T> AddRange(CsLuaList<T> range)
        {
            throw new NotImplementedException();
        }

        public T First()
        {
            return this.list.First();
        }

        public T First(Func<T, bool> condition)
        {
            return this.list.First(condition);
        }

        public T FirstOrDefault()
        {
            return this.list.FirstOrDefault();
        }

        public T FirstOrDefault(Func<T, bool> condition)
        {
            return this.list.FirstOrDefault(condition);
        }

        public T Last()
        {
            return this.list.Last();
        }

        public T Last(Func<T, bool> condition)
        {
            return this.list.Last(condition);
        }

        public T LastOrDefault()
        {
            return this.list.LastOrDefault();
        }

        public T LastOrDefault(Func<T, bool> condition)
        {
            return this.list.LastOrDefault(condition);
        }

        public T Single()
        {
            return this.list.Single();
        }

        public T Single(Func<T, bool> condition)
        {
            return this.list.Single(condition);
        }

        public T SingleOrDefault()
        {
            return this.list.SingleOrDefault();
        }

        public T SingleOrDefault(Func<T, bool> condition)
        {
            return this.list.SingleOrDefault(condition);
        }

        public void Foreach(Action<T> action)
        {
            foreach (var item in this.list)
            {
                action(item);
            }
        }

        public double Max(Func<T, double> selector)
        {
            return this.list.Max(selector);
        }

        public double Min(Func<T, double> selector)
        {
            return this.list.Max(selector);
        }

        public double Sum(Func<T, double> selector)
        {
            return this.list.Sum(selector);
        }

        public CsLuaList<T> OrderBy(Func<T, double> selector)
        {
            return new CsLuaList<T>(this.list.OrderBy(selector).ToList());
        }

        public CsLuaList<T> Union(CsLuaList<T> otherList)
        {
            var newList = this.Where(item => true);
            otherList.Foreach(item => newList.Add(item));
            return newList;
        }

        public CsLuaList<T> Distinct()
        {
            return new CsLuaList<T>(this.list.Distinct().ToList());
        }

        public void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            info.AddValue("__size", this.Count);
            for (var i = 0; i < this.Count; i++)
            {
                var value = new KeyValuePair<object, object>(i, this[i]);
                info.AddValue("" + i, value);
            }
        }
    }
}
